### PR TITLE
Support zeroDateTimeBehavior=convertToNull

### PR DESCRIPTION
**Pull request by tirsen**
This implements zeroDateTimeBehavior=convertToNull like the standard
MySQL driver.

### CODE REVIEW

**Comment by tirsen** [comment]
Fixes https://github.com/youtube/vitess/issues/2135

File: java/jdbc/src/test/java/io/vitess/jdbc/VitessResultSetTest.java [hunk 1]
```diff
@@ -121,12 +123,14 @@ public Cursor getCursorWithRows() {
                 .addLengths("HELLO TDS TEAM".length()).addLengths("HELLO TDS TEAM".length())
                 .addLengths("N".length()).addLengths("HELLO TDS TEAM".length())
                 .addLengths("1".length()).addLengths("val123".length())
-                .addLengths("val123".length()).setValues(ByteString.copyFromUtf8(
-                "-5050-2300023000-100100-100100-1000100024.52100.432016-02-06 " +
-                "14:15:162016-02-0612:34:562016-02-06 14:15:1620161234.56789HELLO TDS TEAMHELLO TDS TEAMHELLO" +
-                " TDS TEAMHELLO TDS TEAMNHELLO TDS TEAM1val123val123"))).build());
+                .addLengths("val123".length()).addLengths("0000-00-00 00:00:00".length())
```
Comment (mpawliszyn) [inline java/jdbc/src/test/java/io/vitess/jdbc/VitessResultSetTest.java]: Do you want two lengths?

Comment (tirsen) [reply java/jdbc/src/test/java/io/vitess/jdbc/VitessResultSetTest.java]: No the val123 was there from before

File: java/jdbc/src/main/java/io/vitess/jdbc/VitessResultSet.java [hunk 2]
```diff
@@ -480,6 +480,13 @@ public Timestamp getTimestamp(int columnIndex) throws SQLException {
             return null;
         }
 
+        Constants.ZeroDateTimeBehavior zeroDateTimeBehavior =
+            vitessStatement.getConnection().getZeroDateTimeBehavior();
+        if (zeroDateTimeBehavior == Constants.ZeroDateTimeBehavior.CONVERTTONULL
+            && this.row.getRawValue(columnIndex).startsWith(Constants.ZERO_DATE_TIME_PREFIX)) {
```
Comment (mpawliszyn) [inline java/jdbc/src/main/java/io/vitess/jdbc/VitessResultSet.java]: StartsWith .. and nothing more

Comment (tirsen) [reply java/jdbc/src/main/java/io/vitess/jdbc/VitessResultSet.java]: Yeah this mirrors the behavior of the mysql jdbc driver:
https://github.com/mysql/mysql-connector-j/blob/3289a357af6d09ecc1a10fd3c26e95183e5790ad/src/com/mysql/jdbc/ResultSetRow.java#L141

**Comment by harshit-gangal** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks broke.

---

*[java/jdbc/src/main/java/io/vitess/jdbc/ConnectionProperties.java, line 340 at r1](https://reviewable.io:443/reviews/youtube/vitess/3178#-KuAJkT_AxQdfZa0QSg9:-KuAJkT_AxQdfZa0QSgA:b5o4dau) ([raw file](https://github.com/youtube/vitess/blob/78d4e3178f78b41656fe91a4f3cb6943e8603990/java/jdbc/src/main/java/io/vitess/jdbc/ConnectionProperties.java#L340)):*
> ```Java
>     }
> 
>     public Constants.ZeroDateTimeBehavior getZeroDateTimeBehavior() {
> ```

This is added in between getter and setter of other field. Can this be moved.

---

*[java/jdbc/src/main/java/io/vitess/jdbc/VitessResultSet.java, line 484 at r1](https://reviewable.io:443/reviews/youtube/vitess/3178#-KuAK3b_Dn2S-KqyCgvD:-KuAK3b_Dn2S-KqyCgvE:b-qvk6uu) ([raw file](https://github.com/youtube/vitess/blob/78d4e3178f78b41656fe91a4f3cb6943e8603990/java/jdbc/src/main/java/io/vitess/jdbc/VitessResultSet.java#L484)):*
> ```Java
> 
>         Constants.ZeroDateTimeBehavior zeroDateTimeBehavior =
>             vitessStatement.getConnection().getZeroDateTimeBehavior();
> ```

we generally use this.vitessStatement ... to know if it is local variable or class member.
good to have.

---

*[java/jdbc/src/main/java/io/vitess/jdbc/VitessResultSet.java, line 486 at r1](https://reviewable.io:443/reviews/youtube/vitess/3178#-KuALBaN95aOQJyB7F_1:-KuAL8H0dAWLsprx2jO8:b-s3c86b) ([raw file](https://github.com/youtube/vitess/blob/78d4e3178f78b41656fe91a4f3cb6943e8603990/java/jdbc/src/main/java/io/vitess/jdbc/VitessResultSet.java#L486)):*
> ```Java
>             vitessStatement.getConnection().getZeroDateTimeBehavior();
>         if (zeroDateTimeBehavior == Constants.ZeroDateTimeBehavior.CONVERTTONULL
>             && this.row.getRawValue(columnIndex).startsWith(Constants.ZERO_DATE_TIME_PREFIX)) {
> ```

This check should be done for getDate and getTime too?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/3178)*

**Comment by tirsen** [comment]
*[java/jdbc/src/main/java/io/vitess/jdbc/VitessResultSet.java, line 486 at r1](https://reviewable.io:443/reviews/youtube/vitess/3178#-KuALBaN95aOQJyB7F_1:-L-b7k8YPvInkkbulDlb:b-wqcvk2) ([raw file](https://github.com/youtube/vitess/blob/78d4e3178f78b41656fe91a4f3cb6943e8603990/java/jdbc/src/main/java/io/vitess/jdbc/VitessResultSet.java#L486)):*
<details><summary><i>Previously, harshit-gangal wrote…</i></summary><blockquote>

This check should be done for getDate and getTime too?
</blockquote></details>

Ah yes on getDate, but not on getTime. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/3178)*

**Comment by tirsen** [comment]
@harshit-gangal Ok, I've addressed your code review comments.

**Comment by harshit-gangal** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/3178#-:-L-bTztDG9H-XDDgo7Kq:bnfp4nl)*

**Comment by harshit-gangal** [comment]
This has conflicts. Kindly resolve.

**Comment by sougou** [comment]
Need conflict resolution and rebase.

**Comment by tirsen** [comment]
@sougou Rebased and conflicts resolve.

**Comment by tirsen** [comment]
I don't understand the test failure. Is it just flaky?

**Comment by sougou** [comment]
Looks like the Java tests need fixing.

**Comment by tirsen** [comment]
I see! You gotta look at the raw log to see the failure... Fixing...

**Comment by sougou** [comment]
LGTM

**Pull request merged**